src/pybullet_blocks/envs/base_env.py

```diff
@@ -427,16 +427,51 @@ def step(
         self,
         action: NDArray[np.float32],
     ) -> tuple[NDArray[np.float32], SupportsFloat, bool, bool, dict[str, Any]]:
-
+        """Take a step in the environment."""
         assert self.action_space.contains(action)
         action_obj = PyBulletBlocksAction.from_vec(action)
 
         # Update robot arm joints.
         joint_arr = np.array(self.robot.get_joint_positions())
         # Assume that first 7 entries are arm.
         joint_arr[:7] += action_obj.robot_arm_joint_delta
+        clipped_joints = np.clip(
+            joint_arr, self.robot.joint_lower_limits, self.robot.joint_upper_limits
+        )
+
+        # Set new robot position.
+        self.robot.set_joints(clipped_joints.tolist())
+
+        # Update held object if exists.
+        if self.current_grasp_transform:
+            world_to_robot = self.robot.get_end_effector_pose()
+            world_to_block = multiply_poses(
+                world_to_robot, self.current_grasp_transform
+            )
+            assert self.current_held_object_id is not None
+            set_pose(
+                self.current_held_object_id,
+                world_to_block,
+                self.physics_client_id,
+            )
+
+        # Get observation after movement.
+        observation = self.get_state().to_observation()
+        info = self._get_info()
+
+        # Check for collisions with table in new state.
+        has_collision = check_body_collisions(
+            self.robot.robot_id,
+            self.table_id,
+            self.physics_client_id,
+            distance_threshold=1e-6,
+        )
 
-        # Update gripper if required.
+        # Return with negative reward if collision.
```

Comment:
I don't think this logic is quite right -- wouldn't we return an observation here where the robot is penetrating the table? also, I think this is now redundant with some of the code in the loop below.